src/blk/kernel/KernelDevice.cc

```diff
@@ -554,28 +554,27 @@ void KernelDevice::_discard_update_threads()
     }
   // Decrease? Signal threads after telling them to stop
   } else if (newcount < oldcount) {
+    std::vector<std::shared_ptr<DiscardThread>> discard_threads_to_stop;
     dout(10) << __func__ << " stopping " << (oldcount - newcount) << " existing discard threads" << dendl;
 
     // Signal the last threads to quit, and stop tracking them
-    for(uint64_t i = oldcount; i > newcount; i--)
-    {
+    for(uint64_t i = oldcount; i > newcount; i--) {
       discard_threads[i-1]->stop = true;
-      discard_threads[i-1]->detach();
+      discard_threads_to_stop.push_back(std::move(discard_threads[i-1]));
```

Comment:
Not sure we really need std::move call here - that's not a hot path so we don't care about the performance much here.
While having that move call might look a bit confusing.
But up to you.

Reply:
Yes, it is not necessary to use move. updated